### PR TITLE
Feat/azure inference support

### PR DESCRIPTION
**Pull request by b4s36t4**
**Title:** 
- Added support to communicate with azure inference API

**Related Issues:** (optional)
- #492

### CODE REVIEW

File: src/providers/azure-ai-inference/complete.ts
Comment (narengogi) [inline src/providers/azure-ai-inference/complete.ts]: gateway appends `provider` key in response for all the providers for enriched logging, you should add that here and for rest of the response transforms

File: src/providers/azure-ai-inference/embed.ts
Comment (narengogi) [inline src/providers/azure-ai-inference/embed.ts]: why would we need to trim input or calculate tokens?

**Review by narengogi — APPROVED** [review]
LGTM, only adding provider field is required

**Comment by b4s36t4** [comment]
**Stream**:

```sh
curl --request POST \
  --url <gateway_url>/v1/chat/completions \
  --header 'Authorization: API_KEY' \
  --header 'content-type: application/json' \
  --header 'x-portkey-azure-api-version: 2024-05-01-preview' \
  --header 'x-portkey-azure-deployment-name: Llama-2-7b-chat-cgnkg' \
  --header 'x-portkey-azure-deployment-type: serverless' \
  --header 'x-portkey-azure-region: eastus' \
  --header 'x-portkey-provider: azure-ai' \
  --data '{
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant"
    },
    {
      "role": "user",
      "content": "Hello?"
    }
  ],
  "stream": true
}'
```

**Non-Stream**
```sh
curl --request POST \
  --url <gateway_url>/v1/chat/completions \
  --header 'Authorization: API_KEY' \
  --header 'content-type: application/json' \
  --header 'x-portkey-azure-api-version: 2024-05-01-preview' \
  --header 'x-portkey-azure-deployment-name: Llama-2-7b-chat-cgnkg' \
  --header 'x-portkey-azure-deployment-type: serverless' \
  --header 'x-portkey-azure-region: eastus' \
  --header 'x-portkey-provider: azure-ai' \
  --data '{
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant"
    },
    {
      "role": "user",
      "content": "Hello?"
    }
  ]
}'
```

**Comment by b4s36t4** [comment]
```sh
curl --request POST \
  --url http://localhost:49454/v1/chat/completions \
  --header 'Authorization: FebCyFH6tU6w59LVhXG37MZFN6mKPYeZ' \
  --header 'content-type: application/json' \
  --header 'x-portkey-provider: github' \
  --data '{
  "max_tokens": 10,
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant"
    },
    {
      "role": "user",
      "content": "Hello?"
    }
  ],
  "model":"AI21-Jamba-Instruct"
}'
```

Github Provider Example.

File: src/handlers/handlerUtils.ts
Comment (VisargD) [inline src/handlers/handlerUtils.ts]: nitpick: Can we rename this to azureApiVersion to keep everything uniform

File: src/providers/azure-ai-inference/api.ts
Comment (b4s36t4) [inline src/providers/azure-ai-inference/api.ts:31]: https://github.com/Azure/azure-sdk-for-python/blob/31688d9ab355f96384dc3b2bdaee84a3b5c7f021/sdk/ml/azure-ai-ml/azure/ai/ml/operations/_local_endpoint_helper.py#L88-L89

File: src/providers/azure-ai-inference/api.ts
Comment (VisargD) [inline src/providers/azure-ai-inference/api.ts]: Can we use the constant mentioned in the globals file for GITHUB

File: src/providers/azure-ai-inference/api.ts
Comment (VisargD) [inline src/providers/azure-ai-inference/api.ts]: I think github does not support azureApiVersion

**Comment by VisargD** [comment]
Can we also add fields in requstValidator schema

**Review by VisargD — APPROVED** [review]

File: src/providers/azure-ai-inference/api.ts
Comment (Tug) [inline src/providers/azure-ai-inference/api.ts:27]: @b4s36t4 This seem to make the azure-ai provider fail at the moment:
```
Error code: 400 - {'detail': "Unexpected EXTRA_PARAMETERS option ignore, expected options are ['error', 'drop', 'pass-through']", 'provider': 'azure-ai'}
```
I tried cloning the project and replacing it with `drop` to see if it fixes it it but the provider isn't available when running the project locally so I'm not sure how to proceed here.

Comment (b4s36t4) [reply src/providers/azure-ai-inference/api.ts:27]: @Tug is it possible to share the request as cURL for me to test in my local. Would need to replicate the issue from my side as well.

Comment (b4s36t4) [reply src/providers/azure-ai-inference/api.ts:27]: Hey, raised a PR https://github.com/Portkey-AI/gateway/pull/901. This should ideally give you the capability of passing the `extra-params` from the headers with request, with default as `pass-through`

You can use `x-portkey-azure-extra-params` header to specify the value for `extra-parameters`, I'll share an update with you once it's merged and released.

Comment (Tug) [reply src/providers/azure-ai-inference/api.ts:27]: > @Tug is it possible to share the request as cURL for me to test in my local. Would need to replicate the issue from my side as well.

Sure it's as simple as
```
curl -X POST "https://api.portkey.ai/v1/chat/completions" \
-H "Content-Type: application/json" \
-H "Authorization: Bearer REDACTED" \
-H "x-portkey-config: REDACTED" \
-H "x-portkey-provider: azure-ai" \
-H "x-portkey-virtual-key: REDACTED" \
-d '{
  "messages": [
    {
      "role": "system",
      "content": "You are a helpful assistant."
    },
    {
      "role": "user",
      "content": "How many languages are in the world?"
    }
  ],
  "temperature": 0,
  "stream": false
}'
```
The virtual key is configured for an azure AI serverless deployment.

Comment (b4s36t4) [reply src/providers/azure-ai-inference/api.ts:27]: Thanks for the response, as I have stated above the PR needs approval, will update you as soon as it landed in the release pipeline.

Comment (Tug) [reply src/providers/azure-ai-inference/api.ts:27]: @b4s36t4 It got merged so I gave it a try and it fixed it, thanks a lot 🙏

**Pull request merged**
